scripts/zkai_d128_block_receipt_composition_gate.py

```
@@ -0,0 +1,1123 @@
+        ("gate/value output", gate["target_commitments"]["gate_value_projection_output_commitment"]),
+        ("hidden activation", activation["target_commitments"]["hidden_activation_commitment"]),
+        ("residual delta", residual["source_commitments"]["residual_delta_commitment"]),
+    ):
+        if commitment == final_output:
+            raise D128BlockReceiptError(f"{label} commitment relabeled as final output")
+
+
+def _validate_receipt(payload: dict[str, Any]) -> None:
+    receipt = payload.get("block_receipt")
+    if not isinstance(receipt, dict):
+        raise D128BlockReceiptError("block receipt must be an object")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)